lib/phoenix_component/declarative.ex

```diff
@@ -398,24 +399,42 @@ defmodule Phoenix.Component.Declarative do
         :ok
     end
 
-    case Atom.to_string(type) do
-      "Elixir." <> _ -> {:struct, type}
-      _ when type in @valid_types -> type
-      _ -> bad_type!(slot, name, type, line, file)
+    cond do
+      type in @valid_types -> type
+      is_tuple(type) -> validate_function_attr_type!(slot, name, type, line, file)
+      type |> Atom.to_string() |> String.starts_with?("Elixir.") -> {:struct, type}
+      true -> bad_type!(slot, name, type, line, file)
     end
   end
 
   defp validate_attr_type!(_module, _key, slot, name, type, line, file) do
     bad_type!(slot, name, type, line, file)
   end
 
+  defp validate_function_attr_type!(slot, name, {:function, args} = type, line, file) do
+    cond do
+      is_integer(args) -> type
+      Keyword.keyword?(args) and Enum.all?(args, fn {_, v} -> v in @valid_types end) -> type
+      is_list(args) and Enum.all?(args, &(&1 in @valid_types)) -> type
+      true -> bad_type!(slot, name, type, line, file)
+    end
+  end
+
+  defp validate_function_attr_type!(slot, name, type, line, file) do
+    bad_type!(slot, name, type, line, file)
+  end
+
   defp bad_type!(slot, name, type, line, file) do
     compile_error!(line, file, """
     invalid type #{inspect(type)} for attr #{attr_slot(name, slot)}. \
     The following types are supported:
 
       * any Elixir struct, such as URI, MyApp.User, etc
       * one of #{Enum.map_join(@builtin_types, ", ", &inspect/1)}
+      * a function written as:
+          * with specific arity, ex: {:function, 2}
```

Comment:
I would only support `{:fun, 2}`. I believe the original idea is that we will mostly support types that we can cheaply check with guards, because we want to eventually emit guards for the patterns (we might already do it in some cases). We cannot check a function receives arguments of a certain shape as guards (that requires a type system and the goal of the types here are not to be a type system either!).

Reply:
In this case, could we add an optional keyword option to the attr that would allow the user to specify clearly the args?

For example, something like this:

```elixir
attr, :some_attribute, {:fun, 2}, args: [arg_1: :string, arg_2: :integer]
```

That way we could do the guard check with `{:fun, 2}`, but still give important information for the user of the component on how they should send their function. Otherwise the main issue still exists where the user will always need to find inside the component implementation code to understand what these two arguments are.

Reply:
I am -1. That’s not the goal of these types. It is not meant to mirror a full type system. For example, lists and tuples are just :list and :tuple. We could maybe add those, but that’s a whole separate discussion. This PR should stick to the current rules.

Reply:
> Otherwise the main issue still exists where the user will always need to find that attr usage inside the component implementation code to understand what these two arguments are.

I’d argue that a good component library would also have good documentation for this. So instead of looking at the code you’d just look at the docs :)

Reply:
> I’d argue that a good component library would also have good documentation for this. So instead of looking at the code you’d just look at the docs :)

That's fair. But at the same time that is not structured, if I have a big library and want to auto generate documentation based on the components code, having that inside the attr call would make it simpler to get the information instead of having to look into a `@doc` string. But that is probably not something that would affect a lot of people I guess.

In any case, I will do the changes suggested by @josevalim so this can be merged supporting `{:fun, 2}` syntax and then I will create a new discussion about expanding that (and probably :list and :tuple). I guess the correct place to do that is the Proposals/Suggestions category in Phoenix forum right?

Reply:
All right, sorry for taking so long to update this PR, but I just pushed the changes that we discussed above.

@josevalim @SteffenDE are the current changes good to be merged?